### DISCUSSION TITLE
🐛 How to use with difftool? IE how to stage changes after reviewing them?

### DISCUSSION THREAD

**Discussion by dessalines** [dandavison/delta#1383] [General]
I'm trying to switch from vimdiff to delta, but I'm having trouble with this simple and common use case (code review) : 

- I've changed a bunch of files.
- I run difftool to see a side-by-side diff of the changes of each file I've changed.
- If I like the changes, I can run `git add` from within the interface in some way.
- I can also see my staged changes if I want, before I run `git commit`

Is there any flow / way to do this using delta?

**Comment by dandavison** [comment]
Hi @dessalines, this question is about using Delta with Vim -- is that correct?

I haven't looked into that myself; others have but I'm not sure what the state of the art is. In case it helps as a reminder -- Delta itself is a command line tool that takes in git/diff/grep output and outputs a reformatted version colorized (incl. syntax highlighting) using ANSI escape sequences.

I'll convert this to a Discussion since AIUI it is Vim-specific rather than a Delta bug/feature request, although if it gives rise to a Delta feature request that's great. (All the things you mention are possible with the [Delta Magit integration](https://dandavison.github.io/delta/tips-and-tricks/using-delta-with-magit.html)).

**Reply by dessalines** [reply]
I'd rather not use vim at all, only delta to review my code changes, and add them to git if everything looks okay. Right now I don't know how to use delta as a difftool.

**Reply by dandavison** [reply]
OK I see. I have to admit that I've never personally looked into the "difftool" functionality of git (personally I use Magit when I want more than git CLI). I know it's come up tangentially in various delta issues. Could you link to something explaining exactly what contract a program must satisfy in order to be used as a "difftool"?

**Reply by dessalines** [reply]
I haven't been able to find any docs from git on how to create one, but it seems like any diff program can work: https://stackoverflow.com/a/6412722/1655478

I'm not sure if that solves the use case of being able to `git add` files after you've reviewed them tho, which is the most important thing for me.

**Reply by dandavison** [reply]
> I'm not sure if that solves the use case of being able to git add files after you've reviewed them tho, which is the most important thing for me.

Just to check -- are you familiar with using Delta with `git add -p`? It's necessary to have

```gitconfig
[interactive]
    diffFilter = delta --color-only
```

as the Delta instructions suggest.

**Reply by dessalines** [reply]
Whoa I didn't even know about that. Is there any way to use that with the side-by-side comparison?

**Reply by dandavison** [reply]
I don't think so -- git requires that the lines displayed by the diff filter are in 1-1 correspondence with the lines of the original (unified) diff.

**Reply by mcepl** [reply]
> > I'm not sure if that solves the use case of being able to git add files after you've reviewed them tho, which is the most important thing for me.
> 
> Just to check -- are you familiar with using Delta with `git add -p`? It's necessary to have

Just for the clarification, this is about using Delta with `git add -p`, not that delta would be able of interactive editing of changes (something `vim -d`), right? Are there any plans of extending Delta in that direction.

If there are not, does anybody know of some `vim -d` alternative, which is not `nvim -d` ;) or `emacs`?
